hs/src/Reach/Simulator/Server.hs

```diff
@@ -0,0 +1,154 @@
+{-# OPTIONS_GHC -Wno-deriving-defaults #-}
+{-# OPTIONS_GHC -Wno-missing-methods #-}
+
+module Reach.Simulator.Server where
+
+import Reach.AST.LL
+import Control.Monad.Reader
+import qualified Reach.Simulator.Core as C
+import Control.Concurrent.STM
+import Data.Default.Class
+import Data.Text.Lazy (Text)
+import Network.Wai.Middleware.RequestLogger
+import Web.Scotty.Trans
+import qualified Data.Map.Strict as M
+-- import GHC.Generics
+-- import Data.Aeson (FromJSON, ToJSON)
+
+instance Default Session where
+  def = initSession
+
+newtype WebM a = WebM { runWebM :: ReaderT (TVar Session) IO a }
+  deriving (Applicative, Functor, Monad, MonadIO, MonadReader (TVar Session))
+
+webM :: MonadTrans t => WebM a -> t WebM a
+webM = lift
+
+gets :: (Session -> b) -> WebM b
+gets f = ask >>= liftIO . readTVarIO >>= return . f
+
+modify :: (Session -> Session) -> WebM ()
+modify f = ask >>= liftIO . atomically . flip modifyTVar' f
+
+type StateId = Int
+type ActionId = Int
+
+portNumber :: Int
+portNumber = 3000
+
+-- state
+data Session = Session
+  {  e_states_actions :: M.Map StateId [ActionId]
+  ,  e_nsid :: Int -- next state id
+  ,  e_naid :: Int -- next action id
+  ,  e_res :: C.DLVal
+  ,  e_ids_actions :: M.Map ActionId C.Action
+  ,  e_states_ks :: M.Map StateId C.PartState
+  }
+
+initSession :: Session
+initSession = Session
+  { e_states_actions = M.empty
+  , e_nsid = 0
+  , e_naid = 0
+  , e_res = C.V_Null
+  , e_ids_actions = M.empty
+  , e_states_ks = M.empty
+  }
+
+initProgSim :: LLProg -> WebM C.PartState
+initProgSim ll = do
+  let initSt = C.initState
+  ps <- return $ C.initApp ll initSt
+  processNewState ps
+
+processNewState :: C.PartState -> WebM C.PartState
+processNewState ps = do
+  sid <- gets e_nsid
+  _ <- case ps of
+    C.PS_Done _ _ -> do
+      _ <- return $ putStrLn "EVAL DONE"
+      registerAction sid C.A_None
+    C.PS_Suspend a _ _ -> registerAction sid a
+  let new_res =
+        case ps of
+          C.PS_Done _ v -> v
+          C.PS_Suspend _ _ _ -> C.V_Null
+  modify $ \ st -> st
+    {e_nsid = sid + 1}
+    {e_res = new_res}
+    {e_states_ks = M.singleton sid ps}
+  return ps
+
+registerAction :: StateId -> C.Action -> WebM ActionId
+registerAction sid act = do
+  aid <- gets e_naid
+  modify $ \ st -> st {e_naid = aid + 1}
+  stacts <- gets e_states_actions
+  idacts <- gets e_ids_actions
+  modify $ \ st -> st {e_ids_actions = M.insert aid act idacts}
+  case M.lookup sid stacts of
+    Nothing -> modify $ \ st -> st {e_states_actions = M.insert sid [aid] stacts }
+    Just acts -> modify $ \ st -> st {e_states_actions = M.insert sid (aid:acts) stacts }
+  return aid
+
+unblockProg :: StateId -> ActionId -> C.DLVal -> WebM ()
+unblockProg sid _aid v = do
+  stks <- gets e_states_ks
+  case M.lookup sid stks of
+    Nothing -> do
+      _ <- return $ putStrLn "previous state not found"
+      return ()
+    Just (C.PS_Suspend _a cst k) -> do
+      let ps = k cst v
```

Comment:
The state may have changed since `cst` was saved, because another action might have been chosen and its effects would be lost if you used `cst`